runtime/src/accounts_db.rs

```diff
@@ -816,40 +884,48 @@ impl AccountsDB {
         reclaims: SlotSlice<AccountInfo>,
         expected_single_dead_slot: Option<Slot>,
         no_dead_slot: bool,
-    ) {
-        if !reclaims.is_empty() {
-            let dead_slots = self.remove_dead_accounts(reclaims, expected_single_dead_slot);
-            if no_dead_slot {
-                assert!(dead_slots.is_empty());
-            } else if let Some(expected_single_dead_slot) = expected_single_dead_slot {
-                assert!(dead_slots.len() <= 1);
-                if dead_slots.len() == 1 {
-                    assert!(dead_slots.contains(&expected_single_dead_slot));
-                }
-            }
-            if !dead_slots.is_empty() {
-                self.process_dead_slots(&dead_slots);
+    ) -> (PurgedAccountSlots, ReclaimedAccounts) {
+        if reclaims.is_empty() {
+            return (HashMap::new(), HashMap::new());
+        }
+
+        let (dead_slots, removed_accounts) =
+            self.remove_dead_accounts(reclaims, expected_single_dead_slot);
+        if no_dead_slot {
+            assert!(dead_slots.is_empty());
+        } else if let Some(expected_single_dead_slot) = expected_single_dead_slot {
+            assert!(dead_slots.len() <= 1);
+            if dead_slots.len() == 1 {
+                assert!(dead_slots.contains(&expected_single_dead_slot));
             }
         }
+        let purged_account_slots = if !dead_slots.is_empty() {
+            self.process_dead_slots(&dead_slots)
+        } else {
+            HashMap::new()
+        };
+        (purged_account_slots, removed_accounts)
```

Comment:
I wonder rustc is clever enough to optimize away these collection construction spanning some call-graph altogether only when this `fn` (= `handle_reclaims`) called from `store_with_hashes`.

I'm a bit worried about the wasted computation if the optimization isn't possible because `store_with_hashes` are the one of code path, which should be the fastest.

I think it's decent guess for lively blockchain with very high tps to overwrite existing accounts over and over (otherwise, some program is leaking accounts very badly...). So, we'll end up creating somewhat large these unused bookkeeping collections.

How about passing around `Option<&mut ReclaimResult>` as an argument for the alternative approach?

Reply:
oh, note that `type ReclaimResult = (AccountSlots, AppendVecOffsets);`.

Reply:
Fixed in latest commit! https://github.com/solana-labs/solana/pull/12606/commits/69092648864a17e4a21d5699a51b5c2b6b83c16e